### PR TITLE
Add new market for Nouns Builder auctions

### PR DESCRIPTION
**Pull request by Yiwen-Gao**
## Motivation

This PR attempts to address issue #231. We tried to start a group bid for a PurpleDAO token, but the transaction failed because Purple is a [Nouns Builder DAO](https://nouns.build/) and isn't currently supported as a valid market. 

## Solution

I added a new market for [Nouns Builder auctions](https://github.com/ourzora/nouns-protocol), since they don't conform to either the Nouns or Zora interfaces.

### CODE REVIEW

File: contracts/market-wrapper/NounsBuilderMarkerWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        return state.settled;
+    }
+
+    /**
+     * @notice Finalize the results of the auction
+     * Parameter auctionId is unused because there's at most one ongoing auction for Nouns Builder DAOs
+     */
+    function finalize(uint256) external override {
+        market.settleAuction();
+    }
```
Comment (Yiwen-Gao) [inline contracts/market-wrapper/NounsBuilderMarkerWrapper.sol:131]: How will this function be used? Should I leave the body empty? None of the PartyBid members will have permissions to settle the current auction.

File: contracts/market-wrapper/NounsBuilderMarkerWrapper.sol [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * @notice Submit bid to Market contract
+     * Parameter auctionId is unused because there's at most one ongoing auction for Nouns Builder DAOs
+     */
+    function bid(uint256, uint256 bidAmount) external override {
+        AuctionState memory state = _getAuctionState();
+        market.createBid{ value: bidAmount }(state.tokenId);
+    }
```
Comment (Yiwen-Gao) [inline contracts/market-wrapper/NounsBuilderMarkerWrapper.sol:113]: Should this be rewritten with the ABI?
```suggestion
    function bid(uint256, uint256 bidAmount) external override {
        AuctionState memory state = _getAuctionState();
        (bool success, bytes memory returnData) = address(market).call{ value: bidAmount }(
            abi.encodeWithSignature("createBid(uint256)", state.tokenId)
        );
        require(success, string(returnData));
    }
```

Comment (Yiwen-Gao) [reply contracts/market-wrapper/NounsBuilderMarkerWrapper.sol:113]: I noticed some of the other market wrappers do that.
https://github.com/PartyDAO/party-protocol/blob/2e5bc3856d4346743e379f9fefcc27a1d9de6c48/contracts/market-wrapper/ZoraMarketWrapper.sol#L91-L97
https://github.com/PartyDAO/party-protocol/blob/2e5bc3856d4346743e379f9fefcc27a1d9de6c48/contracts/market-wrapper/FoundationMarketWrapper.sol#L76-L82

**Comment by Yiwen-Gao** [comment]
I wanted to share this to get initial feedback while I write a few test cases for `mainnet.t.deploy.ts`. I'm happy to change the implementation or add more comments!

**Comment by 0xble** [comment]
> I wanted to share this to get initial feedback while I write a few test cases for `mainnet.t.deploy.ts`. I'm happy to change the implementation or add more comments!
> 
> cc @0xble @merklejerk

Thanks for submitting this! Reviewing now and will respond back soon!

**Comment by 0xble** [comment]
@Yiwen-Gao 

We've built an auction type that can repeated bid on auctions for NFTs from the same collection and same market (e.g. Zora), would this better fit for what you need?

The code for it has already been audited and deployed, but not yet supported on the frontend. We encourage you to take a look at the code [here](https://github.com/PartyDAO/party-protocol/blob/main/contracts/crowdfund/RollingAuctionCrowdfund.sol) and let us know what you think!

**Comment by 0xble** [comment]
@Yiwen-Gao

> @0xble if I understand correctly, I think the issue isn't that the crowdfund isn't rolling but rather that it can't interface with Nouns Builder, which is a new type of market that isn't supported at the moment.

You are right, we'd still need a market wrapper to use. But I think the rolling auction may be helpful in the case if the auction you bid on loses because I assumed in that case you'd want to bid on the next auction the next day instead of having to create another crowdfund.

Anyway, I took a look at your code and the Nouns code you linked. Is [this](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol) the code implementing auctions for Nouns Builder as well as Nouns? I think it is because it is the same implementation code used for the [PurpleDAO auction](https://etherscan.io/address/0x43790fe6bd46b210eb27F01306C1D3546AEB8C1b) your party was interested on bidding on. If I am not missing anything, I think you could use this [Nouns market wrapper](https://github.com/PartyDAO/party-protocol/blob/main/contracts/market-wrapper/NounsMarketWrapper.sol#L29) we've already built for your use case?

Let me know if I am missing anything.

**Comment by Yiwen-Gao** [comment]
@0xble thanks for the feedback!
> you'd want to bid on the next auction the following day instead of having to create another crowdfund

I believe Nouns Builder auctions will [return funds](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol#L168) to the addresses of any losing bidders. Based on my understanding of PartyBid, I think the money would go back to the Party treasury and any member can place a bid on the new auction the next day.

> Is [this](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol) the code implementing auctions for Nouns Builder as well as Nouns?

The contracts for NounsDAO are in [nounsDAO/nouns-monorepo](https://github.com/nounsDAO/nouns-monorepo/tree/master/packages/nouns-contracts) while the contracts for Nouns Builder DAOs are in [ourzora/nouns-protocol](https://github.com/ourzora/nouns-protocol). While the two are quite similar, the function signatures are slightly different, hence why I think we need to implement a new market to support Nouns Builder DAOs. For example, 
- Nouns Builder doesn't implement the `nouns()` function 
- Nouns has `minBidIncrementPercentage()` while Nouns Builder has `minBidIncrement()`
- they have different return types for `auction()`

https://github.com/PartyDAO/party-protocol/blob/2e5bc3856d4346743e379f9fefcc27a1d9de6c48/contracts/vendor/markets/INounsAuctionHouse.sol#L52-L61
https://github.com/PartyDAO/party-protocol/blob/b088c4c6182d1756892fbf94e47cc95fc5c2b9c7/contracts/vendor/markets/INounsBuilderAuctionHouse.sol#L10-L21

**Comment by 0xble** [comment]
> @0xble thanks for the feedback!
> 
> > you'd want to bid on the next auction the following day instead of having to create another crowdfund
> 
> I believe Nouns Builder auctions will [return funds](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol#L168) to the addresses of any losing bidders. Based on my understanding of PartyBid, I think the money would go back to the Party treasury and any member can place a bid on the new auction the next day.
> 
> > Is [this](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol) the code implementing auctions for Nouns Builder as well as Nouns?
> 
> The contracts for NounsDAO are in [nounsDAO/nouns-monorepo](https://github.com/nounsDAO/nouns-monorepo/tree/master/packages/nouns-contracts) while the contracts for Nouns Builder DAOs are in [ourzora/nouns-protocol](https://github.com/ourzora/nouns-protocol). While the two are quite similar, the function signatures are slightly different, hence why I think we need to implement a new market to support Nouns Builder DAOs. For example,
> 
> * Nouns Builder doesn't implement the `nouns()` function
> * Nouns has `minBidIncrementPercentage()` while Nouns Builder has `minBidIncrement()`
> * they have different return types for `auction()`
> 
> https://github.com/PartyDAO/party-protocol/blob/2e5bc3856d4346743e379f9fefcc27a1d9de6c48/contracts/vendor/markets/INounsAuctionHouse.sol#L52-L61
> 
> 
> https://github.com/PartyDAO/party-protocol/blob/b088c4c6182d1756892fbf94e47cc95fc5c2b9c7/contracts/vendor/markets/INounsBuilderAuctionHouse.sol#L10-L21

Okay, that makes sense.

Could you add some test cases for this then and we can review and merge it? I think you could basically just copy [this test file for Nouns market wrapper](https://github.com/PartyDAO/party-protocol-internal/blob/feat/add-ragequit/sol-tests/crowdfund/NounsCrowdfundForked.t.sol#L17) as a template and update it.

File: contracts/market-wrapper/NounsBuilderMarkerWrapper.sol (hunk 1 above)
Comment (Yiwen-Gao) [reply contracts/market-wrapper/NounsBuilderMarkerWrapper.sol]: It seems like this function is only used in tests?

**Comment by Yiwen-Gao** [comment]
@0xble the `NounsCrowdfundForked.t.sol` file helped a lot for writing my tests!

I pushed up the Nouns Builder test file, but I haven't run it locally yet because it seems like the contracts have to be deployed first? The addresses I'm using now are copied from the Zora test file, so they're just a stand-in for now.

https://github.com/PartyDAO/party-protocol/blob/2a3c2a448313ed7a6bc8b9946b755cbdcfd3bd02/sol-tests/crowdfund/NounsBuilderCrowdfundForked.t.sol#L32-L34

**Comment by neokry** [comment]
For nouns builder contracts the reserve price / min bid increment can be changed via DAO vote at any time. This could lead to a potential issue where these values are changed after the wrapper contract is deployed. This is a bit of an edge case but it might make sense to change these from immutable to requested  in getMinimum bid @Yiwen-Gao

**Comment by Yiwen-Gao** [comment]
@neokry good call! I had originally done that to save on contract calls, but since they can be changed, I pushed up a new commit to remove the logic.

File: sol-tests/crowdfund/NounsBuilderCrowdfundForked.t.sol [hunk 3]
```diff
@@ -30,8 +31,8 @@ contract NounsBuilderCrowdfundForkedTest is TestUtils, ERC721Receiver {
     Crowdfund.FixedGovernanceOpts defaultGovOpts;
 
     // Initialize Nouns Builder contracts
-    INounsBuilderAuctionHouse nounsBuilder = INounsBuilderAuctionHouse(0xE468cE99444174Bd3bBBEd09209577d25D1ad673);
-    IMarketWrapper nounsBuilderMarket = IMarketWrapper(0x11c07cE1315a3b92C9755F90cDF40B04b88c5731);
+    INounsBuilderAuctionHouse nounsBuilder = INounsBuilderAuctionHouse(0x43790fe6bd46b210eb27F01306C1D3546AEB8C1b);
```
Comment (Yiwen-Gao) [inline sol-tests/crowdfund/NounsBuilderCrowdfundForked.t.sol:34]: I'm using the address of the PurpleDAO auction house on mainnet.

**Comment by 0xble** [comment]
> I believe Nouns Builder auctions will [return funds](https://github.com/ourzora/nouns-protocol/blob/main/src/auction/Auction.sol#L168) to the addresses of any losing bidders. Based on my understanding of PartyBid, I think the money would go back to the Party treasury and any member can place a bid on the new auction the next day.

Sorry I am just rereading this now and forgot to respond to this part. This is true, however parties are only created after the crowdfund is won (i.e. `Crowdfund` -> `Party`). If during the crowdfund stage it fails to place the highest bid on the PurpleDAO NFT during the one day period the auction is live, the crowdfund would fail and all the funds refunded. Then another crowdfund would have to be created for the next day's auction and re-raise all over again, and it all would have to happen in a single day. As you can probably tell, this is not practical and was the main reason the `RollingAuctionCrowdfund` was built. If I am understanding your usecase correctly, I feel this would be the best auction type for what you're trying to do. You'd use this in combination with this market wrapper you wrote. Happy to walkthrough in more detail how this would be done if I am right about this.

**Comment by Yiwen-Gao** [comment]
> include this in an upcoming audit then merge...you'd might want this sooner?

Yep, I think the sooner the better. But I also understand if we need some time to make sure users' funds will be safe.

> crowdfund would fail and all the funds refunded

Good to know! I haven't looked at the crowdfund code closely, but is this because the auction ID can't be updated? Or because the lifecycle is one way i.e. it can't go from a losing state back to a bidding state again?

> `RollingAuctionCrowdfund`

Nice, what were you thinking for this? Would we just need a test like `RollingNounsCrowdfundForked.t.sol` or additional code changes?

**Comment by Yiwen-Gao** [comment]
@0xble any thoughts on my comments above?

**Comment by 0xble** [comment]
> @0xble any thoughts on my comments above?

Yeah, sorry we've been discussing and planning how to support this internally. Ideally we'd hope to do a deployment and build a frontend for you guys to use this.

Also you are from Juicebox, right? John has considered reaching out to talk about our upcoming release, we have something that may be particularly relevant to your usecase.

**Comment by 0xble** [comment]
> Yep, I think the sooner the better. But I also understand if we need some time to make sure users' funds will be safe.

Okay, are keeping this in mind.

> Good to know! I haven't looked at the crowdfund code closely, but is this because the auction ID can't be updated? Or because the lifecycle is one way i.e. it can't go from a losing state back to a bidding state again?

Yes, this is right. The `RollingAuctionCrowdfund` is able to update to a new auction and token ID after it loses until it either wins or expires.

> Nice, what were you thinking for this? Would we just need a test like `RollingNounsCrowdfundForked.t.sol` or additional code changes?

No new test are needed, since market wrapper share a uniform interface if it works for one market wrapper it should work for all market wrappers in `RollingNounsCrowdfundForked.t.sol`. The new tests we've added are the ones that make sure the wrapper you wrote behaves as expected.

**Comment by Yiwen-Gao** [comment]
> we'd hope to do a deployment and build a frontend for you guys to use this

There's already a frontend at [grouppurple.com](https://www.grouppurple.com/) that users have been using to bid on PurpleDAO tokens. It's currently powered by Juicebox since that's been proven to scale to hundreds of thousands of users. However, JB may be too advanced of a tool for the current state of the project, so we decided to consider alternatives. We had started a custom set of bidding contracts in [Yiwen-Gao/shared-governance-tokens](https://github.com/Yiwen-Gao/shared-governance-tokens) but thought it'd be better to contribute this to the Party Protocol. That way, Nouns Builder auctions can also be available for PartyBid's users and the protocol's other clients.

If we do end up switching the backend to Party Protocol, we can update the website to include something like "powered by Party Protocol" to give credit to the work done by this team? Open to other ways we can integrate as well!

> Also you are from Juicebox, right? 

We don't work for Juicebox, if that's what you mean? 

> John has considered reaching out to talk about our upcoming release, we have something that may be particularly relevant to your usecase

Sounds interesting, always happy to chat!

**Comment by 0xble** [comment]
Hey thanks for your patience. These few weeks have been a particularly busy time with us as because we are about to release a new update to the platform in the coming weeks (which we also think will be of interest based on what you're trying to do 😉). This will be included in our next paid-for audit, so you can be sure it will eventually be included. The main difficulty has just been trying to allow you guys to use it sooner as we assume that is a require for your usecase.

Apologies again for any delay 🙏

**Comment by Yiwen-Gao** [comment]
No worries, thanks for the update!

**Comment by 0xble** [comment]
From what I understood from Danny who you guys talked to your guy's problem has been resolved using the WalletConnect flow for on-chain actions, so I will be closing this PR. Was great working with you guys!

**Pull request closed without merging**